# Part 1

QN 3 -- Import your analysis libraries, as well as your new customer data set as a dataframe.

In [2]:
# Importing Libraries
import pandas as pd
import numpy as np
import os
import matplotlib.pyplot as plt
import seaborn as sns
import scipy

In [3]:
path = r'/Users/stadtsalat/Documents/Instacart Basket Analysis'

In [4]:
# Importing customers.csv
df_customers = pd.read_csv(os.path.join(path, '02 Data', 'Original Data', 'customers.csv'))

In [5]:
# Checking the customers df
df_customers.head()

user_id First Name    Surnam  Gender       STATE  Age date_joined  \
0    26711    Deborah  Esquivel  Female    Missouri   48    1/1/2017   
1    33890   Patricia      Hart  Female  New Mexico   36    1/1/2017   
2    65803    Kenneth    Farley    Male       Idaho   35    1/1/2017   
3   125935   Michelle     Hicks  Female        Iowa   40    1/1/2017   
4   130797        Ann   Gilmore  Female    Maryland   26    1/1/2017   

   n_dependants fam_status  income  
0             3    married  165665  
1             0     single   59285  
2             2    married   99568  
3             0     single   42049  
4             1    married   40374

Qn 4 -- Wrangle the data so that it follows consistent logic; for example, rename columns with illogical names and drop columns that don’t add anything to your analysis.

In [7]:
# Checking the info of the df
df_customers.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 206209 entries, 0 to 206208
Data columns (total 10 columns):
 #   Column        Non-Null Count   Dtype 
---  ------        --------------   ----- 
 0   user_id       206209 non-null  int64 
 1   First Name    194950 non-null  object
 2   Surnam        206209 non-null  object
 3   Gender        206209 non-null  object
 4   STATE         206209 non-null  object
 5   Age           206209 non-null  int64 
 6   date_joined   206209 non-null  object
 7   n_dependants  206209 non-null  int64 
 8   fam_status    206209 non-null  object
 9   income        206209 non-null  int64 
dtypes: int64(4), object(6)
memory usage: 15.7+ MB


In [8]:
# Checking the size of the df
df_customers.shape

(206209, 10)

In [9]:
# Renaming columns
df_customers = df_customers.rename(columns={'First Name': 'first_name','Surnam': 'surname','Gender': 'gender', 'STATE': 'state', 'Age': 'age', 'c, 'n_dependants' : 'number_of_dependants', 'fam_status' : 'marital_status'})


In [10]:
# Checking the column names
df_customers.head(1)

user_id first_name   surname  gender     state  age date_registered  \
0    26711    Deborah  Esquivel  Female  Missouri   48        1/1/2017   

   number_of_dependants marital_status  income  
0                     3        married  165665

 All columns include information that can be helpful. I decided to not drop any of them

QN 5 --Complete the fundamental data quality and consistency checks you’ve learned throughout this Achievement; for example, check for and address missing values and duplicates, and convert any mixed-type data.

Data quality and consistency checks

In [11]:
# Checking for missing values
df_customers.isnull().sum()

user_id                     0
first_name              11259
surname                     0
gender                      0
state                       0
age                         0
date_registered             0
number_of_dependants        0
marital_status              0
income                      0
dtype: int64

There are 11259 misiing values in the first_name column. Since that is a just a bit more than 5% of the total first_name column and we have all the other values, I will not take any further action.

In [12]:
# Checking for duplicates
df_dups = df_customers[df_customers.duplicated()]

In [13]:
df_dups

Empty DataFrame
Columns: [user_id, first_name, surname, gender, state, age, date_registered, number_of_dependants, marital_status, income]
Index: []

There are no duplicates in the df

In [14]:
# Checking for mixed data types
for col in df_customers.columns.tolist():
  weird = (df_customers[[col]].applymap(type) != df_customers[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_customers[weird]) > 0:
    print (col)

first_name


There are mixed data types in the first_name column

In [15]:
# Changing the data type of the first_name column
df_customers['first_name'] = df_customers['first_name'].astype('str')

QN 6 -- Combine your customer data with the rest of your prepared Instacart data. (Hint: Make sure the key columns are the same data type!)

In [16]:
# First we need to import the df
ords_prods_merge = pd.read_pickle(os.path.join(path, '02 Data', 'Prepared Data', 'ords_prods_merge_task48.pkl'))

In [17]:
# Checking the imported df
ords_prods_merge.head(1)

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0   2539329        1             1                   2                  8   

   days_since_prior_order  new_customer  product_id  add_to_cart_order  \
0                     NaN          True         196                  1   

   reordered  ...    price_range_loc     busiest_day        Busiest_days  \
0          0  ...  Mid-range product  Regularly busy  Regulary busy days   

   busiest_period_of_day max_order  loyalty_flag avg_price spending_flag  \
0         Average orders        10  New customer      14.0  High spender   

  median_order   order_frequency_flag  
0         30.0  Non-frequent customer  

[1 rows x 25 columns]

In [18]:
ords_prods_merge.shape

(32404859, 25)

In [19]:
# Merging the two dfs on user_id with an outer join
ords_prods_custs = pd.merge(ords_prods_merge, df_customers, on = 'user_id', how = 'outer')


In [21]:
#Checking the new merged df
ords_prods_custs.shape

(32404859, 34)

In [22]:
#Checking the datatypes of the new merged df
ords_prods_custs.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 32404859 entries, 0 to 32404858
Data columns (total 34 columns):
 #   Column                  Dtype   
---  ------                  -----   
 0   order_id                int64   
 1   user_id                 int64   
 2   order_number            int64   
 3   orders_day_of_week      int64   
 4   order_hour_of_day       int64   
 5   days_since_prior_order  float64 
 6   new_customer            bool    
 7   product_id              int64   
 8   add_to_cart_order       int64   
 9   reordered               int64   
 10  product_name            object  
 11  aisle_id                int64   
 12  department_id           int64   
 13  prices                  float64 
 14  _merge                  category
 15  price_range_loc         object  
 16  busiest_day             object  
 17  Busiest_days            object  
 18  busiest_period_of_day   object  
 19  max_order               int64   
 20  loyalty_flag            object  
 21  avg_pr

In [23]:
# Checking the new df
ords_prods_custs.head(1)

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0   2539329        1             1                   2                  8   

   days_since_prior_order  new_customer  product_id  add_to_cart_order  \
0                     NaN          True         196                  1   

   reordered  ...   order_frequency_flag  first_name  surname  gender  \
0          0  ...  Non-frequent customer       Linda   Nguyen  Female   

     state age date_registered number_of_dependants marital_status  income  
0  Alabama  31       2/17/2019                    3        married   40423  

[1 rows x 34 columns]

The merge was succesful

QN 8 -- Export this new dataframe as a pickle file so you can continue to use it in the second part of this task.

In [28]:
# Exporting to PKL
ords_prods_custs.to_pickle(os.path.join(path, '02 Data','Prepared Data', 'ords_prods_custs.pkl'))
